TW - set job classads to match arch in first and second matchmaking

mapellidario opened the pull request:
Related to #7208 

### Status

tested, to the best of our knowledge so far

- when submitting from lxplus-arm (crab client scram_arch is aarch64), the TW generated jobs with the proper classads. so TW is ok
- however, no arm resource are available in the itb and global pool, so we could make sure that they actually work

### Details

Changes to TW, so that the resulting jobs have the proper classads to match the arch in first and second matchmaking.

belforte commented on src/python/TaskWorker/Actions/DagmanCreator.py:
```diff
@@ -366,12 +366,16 @@ def __init__(self, config, crabserver, procnum=-1, rucioClient=None):
-            if arch == "amd64":
-                info['desired_arch'] = "X86_64"
+            if arch not in SCRAM_TO_ARCH:
+                msg = "Job configured to a ScramArch: '{}' not supported in TaskWorker".format(item)
```
can't you use simple f-format string here ?

mapellidario replied:
sure, I added that line before we decided to use fstrings everywhere, i was trying to be consistent with existing code. but you are right, we should look only towards the future, and use fstrings only :)

belforte commented:
looks good. Please look into the comment above about odd syntax for `msg`
Please also remove at least "12 warnings and errors that must be fixed" from pylint.
The undefined variable is worrying. But I have not looked at what's happening there.

mapellidario commented:
I added two commits. one fixes the bug related to "item" not been defined and added a comments about the regex, the second replaces `.format()` with and fstring and pleases pylint.

belforte commented on src/python/TaskWorker/Actions/DagmanCreator.py:
```diff
@@ -366,12 +366,19 @@ def __init__(self, config, crabserver, procnum=-1, rucioClient=None):
         # Set defaults
-        info['desired_arch'] = "X86_64"
+        info['required_arch'] = "X86_64"
+        # The following regex matches a scram arch into four groups 
```
THANKS ! this explanation was dearly needed

mapellidario replied:
:)

belforte commented on src/python/TaskWorker/Actions/DagmanCreator.py:
```diff
@@ -366,12 +366,19 @@ def __init__(self, config, crabserver, procnum=-1, rucioClient=None):
-            if arch == "amd64":
-                info['desired_arch'] = "X86_64"
+            if arch not in SCRAM_TO_ARCH:
+                msg = f"Job configured to a ScramArch: '{arch}' not supported in TaskWorker"
```
let me be picky, the `as` should not be there. I suggest `to a ScramArch : not...` -->` for non-supporterd ScramArch`

belforte commented on src/python/TaskWorker/Actions/DagmanCreator.py:
```diff
@@ -788,11 +795,11 @@ def createSubdag(self, splitterResult, **kwargs):
         else:
             siteinfo = {'group_sites': {}, 'group_datasites': {}}
-        if os.path.exists("site.ad"):
+        if os.path.exists("site.ad", encoding='utf-8'):
```
move `encoding` to next line !

mapellidario replied:
oops

mapellidario commented:
Thanks stefano! It should be better now

(i eventually installed pylint on my system, not to miss bad errors again!)

belforte commented:
> (i eventually installed pylint on my system, not to miss bad errors again!)

+1
that helps for sure, even before committing to your branch
But maybe we can also find a a way to run Jenkins w/o a PR to master ?

belforte commented:
can you kindly take care of this as well ?

 W0707, line 1195 in DagmanCreator.executeInternal: Consider explicitly re-raising using the 'from' keyword 

I like it better when there is no "warnings and errors that must be fixed"

mapellidario commented:
> can you kindly take care of this as well ?

sure, i did not find it in the table at #7756 and thought you decided to ignore it. the is straightforward, i will edit the table myself

belforte commented:
I only put in that table those which I had to fix now. Did not try to remember everything else. Please keep adding to the table

mapellidario commented:
I see, thanks. 

can we merge this?
